etor/src/main/java/gov/hhs/cdc/trustedintermediary/etor/metadata/PartnerMetadataOrchestrator.java

```diff
@@ -67,19 +99,23 @@ public void updateMetadataForSentOrder(String receivedSubmissionId, String sentS
                     "Unable to retrieve metadata from RS history API", e);
         }
 
+        logger.logInfo("Updating metadata with receiver: {}", receiver);
         partnerMetadata = partnerMetadata.withReceiver(receiver);
         partnerMetadataStorage.saveMetadata(partnerMetadata);
     }
 
     public Optional<PartnerMetadata> getMetadata(String receivedSubmissionId)
             throws PartnerMetadataException {
-        // call the metadata storage to get the metadata.
-        // check if the receiver is filled out, and if it isn't, call the RS history API to get the
-        // receiver.
-        // if had to call the history API, extract the receiver and call the metadata storage to
-        // save the metadata with the receiver added.
-        // return the metadata.
-        return partnerMetadataStorage.readMetadata(receivedSubmissionId);
+        PartnerMetadata partnerMetadata =
+                partnerMetadataStorage.readMetadata(receivedSubmissionId).orElseThrow();
+
+        if (partnerMetadata.receiver() == null && partnerMetadata.sentSubmissionId() != null) {
+            logger.logInfo("Receiver name not found in metadata, looking up from RS history API");
+            updateMetadataForSentOrder(receivedSubmissionId, partnerMetadata.sentSubmissionId());
+            return partnerMetadataStorage.readMetadata(receivedSubmissionId);
+        }
+
+        return Optional.of(partnerMetadata);
```

Comment:
Does this ever return an Optional.empty? If not should it still return an optional?

Reply:
Yes, it should return Optional.empty if the stored metadata was not found